### ISSUE TITLE
TreeExplainer support for balance ensembles from imblearn 

### ISSUE THREAD

**Issue by mrastgoo** [shap/shap#338]
Hi, 
I noticed that the TreeExplainer only supports sklearn, xgboost, lightgbm, and catboost models, is there anyway to introduce a Tree ensemble other than the supported models ? 

I am currently using balanced ensembles (`BalanceRandomForestClassifier`) from `imbalanced-learn` and I noticed that they are using similar API than sklearn. Would it be interesting for you to include them in the list of acceptable tree models in the future ? 

thanks in advance,

**Comment by glemaitre** [comment]
Looking at the code [here](https://github.com/slundberg/shap/blob/master/shap/explainers/tree.py#L392), it would be better to ducktype instead of checking the module strictly which could be a bit brittle if the model is renamed or the location of the module change. At first glance, it should be possible to do something like (only for the beggining of function):

```python
if hasattr(model, 'tree_') or hasattr(model, 'estimators_'):
    self.tree_output = 'probability' if model._estimator_type == 'classifier' else 'raw_value'
    self.objective = objective_name_map.get(model.criterion, None)
    if hasattr(model, 'tree_'):
        self.trees = [Tree(model.tree_)]
    else:
        normalize = True
        # overwrite for Gradient Boosting model
        if hasattr(model, 'learning_rate'):
            normalize = False
            scaling = model.learning_rate
            self.tree_output = 'log_odds' if model._estimator_type == 'classifier' else 'raw_value'
            try:
                 self.base_offset = model.init_.mean if model._estimator_type == 'regressor' else model.init_.prior
            except AttributeError:
                raise ValueError('Unsupported init model ...')
        self.trees = [Tree(est.tree_, scaling=scaling, normalize=normalize) for est in model.estimators_]
elif hasattr(model, 'get_booster') or hasattr(model, 'boost'):
     self.model_type = 'xgboost'
     self.base_offset = None
     json_trees = get_xgboost_json(self.original_model)
     self.trees = [Tree(json.loads(t)) for t in json_trees]
     if hasattr(model, 'get_booster'):
        self.original_model = model.get_booster()
        objective_ = model.objective
     else:
        self.original_model = model
        objective_ = model.attr("objective")
     if objective_ is not None:
         self.objective = objective_name_map.get(objective_, None)
         self.tree_output = tree_output_name_map.get(objective_, None)
elif hasattr(model, 'booster_') or hasattr(model, '__boost'):
    # implementation of lightgbm
...
```

**Comment by slundberg** [comment]
@mrastgoo I am happy to include new supported tree types, they have been added so far as people have brought them up. I'll merge a PR whenever, and when I have time I add them myself.

@glemaitre Thanks for writing up how to duck type it! ...I am a little nervous though that by duck typing we might accidentally make the wrong assumptions for a model type we haven't seen. For example perhaps a classifier that has a learning rate but doesn't use a logistic link function. It seems like there is a tradeoff between two problems: (1) having to patch in a new module path in the future, vs. (2) people getting invalid results without an obvious error. I would prefer to make the first mistake over the second. I am happy to be convinced though if you think I missed something.

Either way we should support the BalanceRandomForestClassifier.

**Comment by mrastgoo** [comment]
@slundberg  thanks for the reply I will try to submit a PR on that, if it is not in process actually.

**Comment by slundberg** [comment]
@mrastgoo great! Just coordinate with @glemaitre to make sure you are on the same page with how it should be done and aren't duplicating any efforts.

**Comment by camilaagw** [comment]
@slundberg and @glemaitre Have you already submitted your pull request? If not I can help :)

**Comment by HurricaneHarriet** [comment]
@camilaagw @slundberg @mrastgoo Hi, has imblearn support been added yet? I'm using imblearn.ensemble.EasyEnsembleClassifier. Thank you!
